src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -533,6 +576,98 @@
             (catch RaiseException e
               (handle-task-details-jruby-exception e environment module task))))))
 
+(schema/defn ^:always-validate
+  all-plans-response! :- ringutils/RingResponse
+  "Process the info, returning a Ring response to be propagated back up to the
+  caller of the endpoint.
+  Returns environment as a list of objects for an eventual future in which
+  tasks for all environments can be requested, and a given task will list all
+  the environments it is found in."
+  [info-from-jruby :- [{schema/Any schema/Any}]
+   environment :- schema/Str
+   jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)]
+  (let [format-plan (fn [plan-object]
+                      {:name (:name plan-object)
+                       :environment [{:name environment
+                                      :code_id nil}]})]
+    (->> (map format-plan info-from-jruby)
+         (middleware-utils/json-response 200))))
+
+(schema/defn ^:alwaysa-validate
+  all-plans-fn :- IFn
+  "Middleware function for constructing a Ring response from an incoming
+  request for plans information."
+  [jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)]
+  (fn [request]
+    (let [environment (jruby-request/get-environment-from-request request)]
+      (if-let [plan-info-for-env
+               (sort-nested-info-maps
+                 (jruby-protocol/get-plans jruby-service
+                                           (:jruby-instance
+                                             request)
+                                           environment))]
+        (all-plans-response! plan-info-for-env
+                             environment
+                             jruby-service)
+        (environment-not-found environment)))))
```

Comment:
What happens when an environment exists but has no plans? what is returned? should it be something other than 200 if there are no plans ?

Reply:
Nope. This is consumed by code, so saying "We got a thing with no items" is much easier for other services to handle than something like a 404

Reply:
Okay, returning 200 makes sense, do we still need to capture the 'environment exists but no plans exist' situation? or do we just return an empty response?

Reply:
We just get an empty list from the info-service and will do all our transformations on each element of that list, so it'll all just work :)